columnflow/production/cms/dy.py

```diff
@@ -0,0 +1,459 @@
+        "mass": np.zeros_like(events.PuppiMET.pt),
+    })
+
+    # Build full and visible boson vectors from generator-level information
+    full = vector.array({
+        "pt": events.gen_dilepton_all.pt,
+        "phi": events.gen_dilepton_all.phi,
+        "eta": np.zeros_like(events.gen_dilepton_all.pt),
+        "mass": np.zeros_like(events.gen_dilepton_all.pt),
+    })
+    vis = vector.array({
+        "pt": events.gen_dilepton_vis.pt,
+        "phi": events.gen_dilepton_vis.phi,
+        "eta": np.zeros_like(events.gen_dilepton_vis.pt),
+        "mass": np.zeros_like(events.gen_dilepton_vis.pt),
+    })
+
+    # Compute the recoil vector U = MET + vis - full
+    Ux = met.x + vis.x - full.x
+    Uy = met.y + vis.y - full.y
+
+    # Project U onto the full boson direction
+    full_pt = full.pt
+    full_unit_x = full.x / full_pt
+    full_unit_y = full.y / full_pt
+    upara = Ux * full_unit_x + Uy * full_unit_y
+    uperp = -Ux * full_unit_y + Uy * full_unit_x
+
+    # Determine jet multiplicity for the event (jet selection as in original)
+    jet_selection = ((events.Jet.pt > 30) & (np.abs(events.Jet.eta) < 2.5)) | (
+        (events.Jet.pt > 50) & (np.abs(events.Jet.eta) >= 2.5)
+    )
+    selected_jets = events.Jet[jet_selection]
+    njet = np.asarray(ak.num(selected_jets, axis=1), dtype=np.float32)
+
+    # Apply nominal recoil correction on U components
+    # (see here: https://cms-higgs-leprare.docs.cern.ch/htt-common/V_recoil/#example-snippet)
+    upara_corr = self.recoil_corrector.evaluate(
+        self.dy_recoil_config.era,
+        self.dy_recoil_config.order,
+        njet,
+        events.gen_dilepton_all.pt,
+        "Upara",
+        upara,
+    )
+    uperp_corr = self.recoil_corrector.evaluate(
+        self.dy_recoil_config.era,
+        self.dy_recoil_config.order,
+        njet,
+        events.gen_dilepton_all.pt,
+        "Uperp",
+        uperp,
+    )
+
+    # Reassemble the corrected U vector
+    U_corr_x = upara_corr * full_unit_x - uperp_corr * full_unit_y
+    U_corr_y = upara_corr * full_unit_y + uperp_corr * full_unit_x
+
+    # Recompute corrected MET: MET_corr = U_corr - vis + full
+    met_corr_x = U_corr_x - vis.x + full.x
+    met_corr_y = U_corr_y - vis.y + full.y
+    met_corr_pt = np.sqrt(met_corr_x**2 + met_corr_y**2)
+    met_corr_phi = np.arctan2(met_corr_y, met_corr_x)
+
+    events = set_ak_column(
+        events, "RecoilCorrMET.pt", met_corr_pt, value_type=np.float32,
+    )
+    events = set_ak_column(
+        events, "RecoilCorrMET.phi", met_corr_phi, value_type=np.float32,
+    )
+
+    # --- Systematic variations ---
+    # Derive H from the nominal corrected MET: H = - (MET_corr + vis)
+    Hx = -met_corr_x - vis.x
+    Hy = -met_corr_y - vis.y
+    H_pt = np.sqrt(Hx**2 + Hy**2)
+    H_phi = np.arctan2(Hy, Hx)
+    # Project H into the full boson coordinate system
+    H_para = H_pt * np.cos(H_phi - full.phi)
+    H_perp = H_pt * np.sin(H_phi - full.phi)
+
```

Comment:
Could you go for lower case variables + snake-case here?

Reply:
Fixed.